### ISSUE TITLE
[Scala] Highlighting lambda parameters

### ISSUE THREAD

**Issue by djspiewak** [sublimehq/Packages#558]
I'm honestly not certain if there even is an efficient way of doing this.  I'd like to be able to highlight lambda parameters as `variable.parameter`.  Unfortunately, the syntax is fairly… diverse.

```scala
{ a => ... }
{ (a, b) => ... }
{ a: Int => ... }
{ (a: Int, b: Int) => ... }

(a) => ...
(a, b) => ...
(a: Int) => ...
(a: Int, b: Int) => ...

a => ...
a: Int => ...
```

`a` and `b` are lambda parameters in all of the above examples.  The ellipses may be replaced by any expression (including other lambdas).

Basically, the marker token here is the `=>`, but we need to backtrack, perhaps quite a distance, in order to find the parameters.  That… doesn't sound like a recipe for efficiency.

One thing that might help here is I have a mostly-finished implementation of type expression parsing on my local fork, which eliminates the need to worry about `=>` appearing in types, and the pattern match parsing already eliminates the need to worry about it in `case` expressions.

**Comment by wbond** [comment]
As long as there isn't a line break between the beginning of the parameters and the `=>`, we can handle it no problem. We tokenize one line at a time for efficiency sake, and to allow for resumable operation. This means we don't have to re-tokenize the entire file when you add a single character on line 200 of a 10,000 line file.

In terms of performance, Jon has written a very fast, custom built regex engine for syntax highlighting. It does not use backtracking, and thus is very fast – It can highlight the 6.6MB `sqlite.c` in around 900ms on my Mac. The contexts and matches required for handling this will likely be simplerthan what we use to try and handle C++ with C preprocessor macros thrown in.

**Comment by djspiewak** [comment]
There *can* be a newline which separates the parameters and the `=>`.  For example:

```scala
{
(a: Int)
=>
42
}
```

That's a valid lambda.  But it's *very* unusual and would be considered bad style by almost everyone.  I'm completely ok with saying that lambdas defined in this way just don't get syntax highlighted.

What would this kind of functionality look like?  I've taken a crack at it in the past, and it's always fallen short in a few cases.

**Comment by wbond** [comment]
I would start with the version JavaScript uses:

 - https://github.com/sublimehq/Packages/blob/master/JavaScript/JavaScript.sublime-syntax#L16
 - https://github.com/sublimehq/Packages/blob/master/JavaScript/JavaScript.sublime-syntax#L781-L784
 - https://github.com/sublimehq/Packages/blob/master/JavaScript/JavaScript.sublime-syntax#L865-L882
 - https://github.com/sublimehq/Packages/blob/master/JavaScript/JavaScript.sublime-syntax#L922-L943

**Comment by wbond** [comment]
I believe this was addressed in #563
